ludwig/datasets/base_dataset.py

```diff
@@ -0,0 +1,42 @@
+import abc
+from typing import Dict
+import pandas as pd
+
+
+"""An abstract base class that defines a set of methods for download,
+preprocess and plug data into the ludwig training API"""
+
+
+class BaseDataset(metaclass=abc.ABCMeta):
+
+    """Download the raw data to the ludwig cache in the format ~/.ludwig_cache/id
+    where is is represented by the name.version of the dataset
+    args:
+        self (BaseDataset): A pointer to the current class
+        dataset_name: The name of the dataset we need to retrieve
+    """
+    @abc.abstractmethod
+    def download(self, dataset_name) -> None:
+        raise NotImplementedError("You will need to implement the download method to download the training data")
+
+    """Process the dataset to get it ready to be plugged into a dataframe
```

Comment:
Instead of loading this into memory, with this function we want to map the raw dataset to the processed dataset in persistent storage.

So:

```
.ludwig_cache/id/raw -> .ludwig_cache/id/processed
```

For some datasets like CSV, this could be a no-op, but if the dataset is in a weird format, or needs to be sanitized in some way, this is where we would do it.

Reply:
Fixed